simphony/cuds/model.py

```diff
@@ -39,119 +55,179 @@ def __init__(self):
 
     @staticmethod
     def _is_dataset(obj):
         """Check if the object is a dataset."""
-        return isinstance(obj, (ABCParticles, ABCLattice, ABCMesh))
+        return isinstance(obj, (ABCParticles,
+                                ABCLattice,
+                                ABCMesh,
+                                api.DataSet))
 
     @property
     def data(self):
         """Data container for CUDS items."""
         return DataContainer(self._data)
 
     @data.setter
-    def data_setter(self, value):
+    def data(self, value):
         self._data = DataContainer(value)
 
+    @data.deleter
+    def data(self):
+        raise TypeError("Can't delete data.")
+
     def add(self, component):
         """Add a component to the CUDS computational model.
 
-        If the component has `uid` attribute but that value is
-        None, a new uuid will be created and assigned to the
-        component after the component is successfully added
-        to the CUDS computational model.
+        This will replace any existing value with the same `uid`
+        or `name`.
 
         Parameters
         ----------
         component: CUDSComponent
-            a component of CUDS, reflecting SimPhoNy metadata
+            a CUDS component according to the SimPhoNy metadata
 
         Raises
         ------
         TypeError
-            if component is not a CUDS component
-
+            if the component is not a CUDS component
         ValueError
-            if the component is already added
+            if a component already exists
         """
-        # Check if the object is valid
-        if not self._is_cuds_component(component):
-            raise TypeError('Not a CUDS component.')
-
-        # FIXME: what if component.uuid was defined before adding the
-        # component and it is not an instance of uuid.UUID,
-        # should we check for it, somewhere in this function?
-
-        try:
-            component_id = component.uuid
-        except AttributeError:
-            # Datasets (ABCParticles, ABCLattice, ABCMesh) do not
-            # have a uid attibute
-            component_id = component.name
-        else:
-            # if the uuid is not defined, create a new one here
-            if component_id is None:
-                component_id = uuid.uuid4()
+        # Only accept CUDSComponent subclasses and datasets
+        if not (isinstance(component, api.CUDSComponent) or
+                self._is_dataset(component)):
+            raise TypeError('Not a CUDSComponent nor a dataset object: %s.' %
+                            type(component))
+
+        # Do not accept items with duplicate names.
+        # Components/datasets with no name will be added, however
+        # it is not possible to get/remove them with `name`
+        if component.name in self._name_to_id_map:
+            raise ValueError('Name clash. A %s component has the same name %s'
+                             % (type(self.get(component.name)),
```

Comment:
I would skip `self.get(component.name)` here to avoid confusing error being raised from `self.get` etc.